## Preprocessing

In [1]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [2]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
application_df = application_df.drop(columns = ['EIN', 'NAME'])
application_df

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0                  T10       Independent          C1000    ProductDev   
1                   T3       Independent          C2000  Preservation   
2                   T5  CompanySponsored          C3000    ProductDev   
3                   T3  CompanySponsored          C2000  Preservation   
4                   T3       Independent          C1000     Heathcare   
...                ...               ...            ...           ...   
34294               T4       Independent          C1000    ProductDev   
34295               T4  CompanySponsored          C3000    ProductDev   
34296               T3  CompanySponsored          C2000  Preservation   
34297               T5       Independent          C3000    ProductDev   
34298               T3       Independent          C1000  Preservation   

       ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS   ASK_AMT  \
0       Association       1              0                      N      5000   
1      Co-operative       1         1-9999                      N    108590   
2       Association       1              0                      N      5000   
3             Trust       1    10000-24999                      N      6692   
4             Trust       1  100000-499999                      N    142590   
...             ...     ...            ...                    ...       ...   
34294   Association       1              0                      N      5000   
34295   Association       1              0                      N      5000   
34296   Association       1              0                      N      5000   
34297   Association       1              0                      N      5000   
34298  Co-operative       1          1M-5M                      N  36500179   

       IS_SUCCESSFUL  
0                  1  
1                  1  
2                  0  
3                  1  
4                  1  
...              ...  
34294              0  
34295              0  
34296              0  
34297              1  
34298              0  

[34299 rows x 10 columns]

In [4]:
# Determine the number of unique values in each column.
application_df.nunique()

APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64

In [5]:
# Look at APPLICATION_TYPE value counts for binning
type_counts = application_df['APPLICATION_TYPE'].value_counts()
type_counts

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [6]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
application_types_to_replace = list(type_counts[type_counts < 500].index)

# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
application_df['APPLICATION_TYPE'].value_counts()

T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: APPLICATION_TYPE, dtype: int64

In [7]:
# Look at CLASSIFICATION value counts for binning
class_counts = application_df['CLASSIFICATION'].value_counts()
class_counts

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [8]:
# You may find it helpful to look at CLASSIFICATION value counts >1
class_counts_gt1 = class_counts.loc[class_counts > 1]
class_counts_gt1

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1234        2
C1246        2
C1267        2
C1256        2
Name: CLASSIFICATION, dtype: int64

In [9]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
classifications_to_replace = list(class_counts[class_counts < 900].index)

# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")

# Check to make sure binning was successful
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: CLASSIFICATION, dtype: int64

In [12]:
# Convert categorical data to numeric with `pd.get_dummies`
numerical_application = pd.get_dummies(application_df)
numerical_application

STATUS   ASK_AMT  IS_SUCCESSFUL  APPLICATION_TYPE_Other  \
0           1      5000              1                       0   
1           1    108590              1                       0   
2           1      5000              0                       0   
3           1      6692              1                       0   
4           1    142590              1                       0   
...       ...       ...            ...                     ...   
34294       1      5000              0                       0   
34295       1      5000              0                       0   
34296       1      5000              0                       0   
34297       1      5000              1                       0   
34298       1  36500179              0                       0   

       APPLICATION_TYPE_T10  APPLICATION_TYPE_T19  APPLICATION_TYPE_T3  \
0                         1                     0                    0   
1                         0                     0                    1   
2                         0                     0                    0   
3                         0                     0                    1   
4                         0                     0                    1   
...                     ...                   ...                  ...   
34294                     0                     0                    0   
34295                     0                     0                    0   
34296                     0                     0                    1   
34297                     0                     0                    0   
34298                     0                     0                    1   

       APPLICATION_TYPE_T4  APPLICATION_TYPE_T5  APPLICATION_TYPE_T6  ...  \
0                        0                    0                    0  ...   
1                        0                    0                    0  ...   
2                        0                    1                    0  ...   
3                        0                    0                    0  ...   
4                        0                    0                    0  ...   
...                    ...                  ...                  ...  ...   
34294                    1                    0                    0  ...   
34295                    1                    0                    0  ...   
34296                    0                    0                    0  ...   
34297                    0                    1                    0  ...   
34298                    0                    0                    0  ...   

       INCOME_AMT_1-9999  INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  \
0                      0                       0                         0   
1                      1                       0                         0   
2                      0                       0                         0   
3                      0                       1                         0   
4                      0                       0                         1   
...                  ...                     ...                       ...   
34294                  0                       0                         0   
34295                  0                       0                         0   
34296                  0                       0                         0   
34297                  0                       0                         0   
34298                  0                       0                         0   

       INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  \
0                       0                 0                       0   
1                       0                 0                       0   
2                       0                 0                       0   
3                       0                 0                       0   
4                       0                 0                       0   
...                   ...            

In [13]:
# Split our preprocessed data into our features and target arrays
X = numerical_application.drop(['IS_SUCCESSFUL'], axis =1)
y = numerical_application['IS_SUCCESSFUL']

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 50)

In [14]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Compile, Train and Evaluate the Model

In [15]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_input_features = len(X_train_scaled[0])
hidden_nodes_layer1 = 10
hidden_nodes_layer2 = 5

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units = hidden_nodes_layer1,
        input_dim = number_input_features, activation = "tanh"))

# Second hidden layer
nn.add(tf.keras.layers.Dense(
    units = hidden_nodes_layer2, activation = "relu"))

# Output layer
nn.add(tf.keras.layers.Dense(units = 1, activation = "relu"))

# Check the structure of the model
nn.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 10)                440       
                                                                 
 dense_1 (Dense)             (None, 5)                 55        
                                                                 
 dense_2 (Dense)             (None, 1)                 6         
                                                                 
Total params: 501
Trainable params: 501
Non-trainable params: 0
_________________________________________________________________


In [16]:
# Compile the model
nn.compile(loss = "binary_crossentropy", optimizer = "adam", metrics =  ["accuracy"])

In [17]:
# Train the model
fit_model = nn.fit(X_train_scaled, y_train, epochs = 100)

Epoch 1/100
804/804 [==============================] - 3s 2ms/step - loss: 3.1291 - accuracy: 0.5415
Epoch 2/100
804/804 [==============================] - 2s 2ms/step - loss: 0.9127 - accuracy: 0.6493
Epoch 3/100
804/804 [==============================] - 2s 3ms/step - loss: 0.8629 - accuracy: 0.6730
Epoch 4/100
804/804 [==============================] - 2s 2ms/step - loss: 0.8541 - accuracy: 0.6852
Epoch 5/100
804/804 [==============================] - 2s 2ms/step - loss: 0.8079 - accuracy: 0.6908
Epoch 6/100
804/804 [==============================] - 2s 2ms/step - loss: 0.7967 - accuracy: 0.6955
Epoch 7/100
804/804 [==============================] - 2s 2ms/step - loss: 0.7917 - accuracy: 0.7037
Epoch 8/100
804/804 [==============================] - 2s 2ms/step - loss: 0.7889 - accuracy: 0.7075
Epoch 9/100
804/804 [==============================] - 2s 2ms/step - loss: 0.7953 - accuracy: 0.7116
Epoch 10/100
804/804 [==============================] - 2s 3ms/step - loss: 0.7940 - accura

In [18]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.5558 - accuracy: 0.7371 - 537ms/epoch - 2ms/step
Loss: 0.5557925701141357, Accuracy: 0.7371428608894348


Results:

Applicaton_Type cutoff = 500
Classification cutoff = 900
layer1 = 10
layer2 = 5
Loss = 0.55579257011413557
Accuracy = 0.7371428608894348

Attempt #2

In [29]:
number_input_features = len(X_train_scaled[0])
hidden_nodes_layer1 = 15
hidden_nodes_layer2 = 10

nn_1 = tf.keras.models.Sequential()

nn_1.add(tf.keras.layers.Dense(units = hidden_nodes_layer1,
        input_dim = number_input_features, activation = "tanh"))

nn_1.add(tf.keras.layers.Dense(
    units = hidden_nodes_layer2, activation = "relu"))

nn_1.add(tf.keras.layers.Dense(units = 1, activation = "relu"))

nn_1.summary()

Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_6 (Dense)             (None, 15)                660       
                                                                 
 dense_7 (Dense)             (None, 10)                160       
                                                                 
 dense_8 (Dense)             (None, 1)                 11        
                                                                 
Total params: 831
Trainable params: 831
Non-trainable params: 0
_________________________________________________________________


In [30]:
nn_1.compile(loss ="binary_crossentropy", optimizer = "adam", metrics = ["accuracy"])

In [32]:
fit_model = nn_1.fit(X_train_scaled, y_train, epochs = 100)

Epoch 1/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5648 - accuracy: 0.7286
Epoch 2/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5608 - accuracy: 0.7255
Epoch 3/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5572 - accuracy: 0.7269
Epoch 4/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5619 - accuracy: 0.7230
Epoch 5/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5571 - accuracy: 0.7280
Epoch 6/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5569 - accuracy: 0.7287
Epoch 7/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5622 - accuracy: 0.7260
Epoch 8/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5639 - accuracy: 0.7240
Epoch 9/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5574 - accuracy: 0.7278
Epoch 10/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5549 - accura

In [33]:
model_loss, model_accuracy = nn_1.evaluate(X_test_scaled, y_test, verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.5536 - accuracy: 0.7374 - 487ms/epoch - 2ms/step
Loss: 0.5535508990287781, Accuracy: 0.7373760938644409


Reults for Atempt 2:

Application cutoff = 500
Classification cutoff = 900
Layer1 = 15
Layer2 = 10
Loss = 0.55355
Accuracy = 0.737376

Attempt 3

In [34]:
number_input_features = len(X_train_scaled[0])
hidden_nodes_layer1 = 50
hidden_nodes_layer2 = 20

nn_2 = tf.keras.models.Sequential()

nn_2.add(tf.keras.layers.Dense(units = hidden_nodes_layer1,
          input_dim = number_input_features, activation = "tanh"))

nn_2.add(tf.keras.layers.Dense(
    units = hidden_nodes_layer2, activation = "relu"))

nn_2.add(tf.keras.layers.Dense(units =1, activation = "relu"))

nn_2.summary()

Model: "sequential_3"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_9 (Dense)             (None, 50)                2200      
                                                                 
 dense_10 (Dense)            (None, 20)                1020      
                                                                 
 dense_11 (Dense)            (None, 1)                 21        
                                                                 
Total params: 3,241
Trainable params: 3,241
Non-trainable params: 0
_________________________________________________________________


In [35]:
nn_2.compile(loss = "binary_crossentropy", optimizer = "adam", metrics = ["accuracy"])

In [36]:
fit_model = nn_2.fit(X_train_scaled, y_train, epochs = 100)

Epoch 1/100
804/804 [==============================] - 2s 2ms/step - loss: 1.2058 - accuracy: 0.6547
Epoch 2/100
804/804 [==============================] - 2s 2ms/step - loss: 0.8453 - accuracy: 0.7037
Epoch 3/100
804/804 [==============================] - 2s 2ms/step - loss: 0.6930 - accuracy: 0.7077
Epoch 4/100
804/804 [==============================] - 2s 3ms/step - loss: 0.7060 - accuracy: 0.7085
Epoch 5/100
804/804 [==============================] - 2s 2ms/step - loss: 0.6247 - accuracy: 0.7184
Epoch 6/100
804/804 [==============================] - 2s 2ms/step - loss: 0.7084 - accuracy: 0.7175
Epoch 7/100
804/804 [==============================] - 2s 2ms/step - loss: 0.9700 - accuracy: 0.6896
Epoch 8/100
804/804 [==============================] - 2s 2ms/step - loss: 0.6761 - accuracy: 0.7165
Epoch 9/100
804/804 [==============================] - 2s 2ms/step - loss: 0.8775 - accuracy: 0.6994
Epoch 10/100
804/804 [==============================] - 2s 2ms/step - loss: 0.7404 - accura

In [37]:
model_loss, model_accuracy = nn_2.evaluate(X_test_scaled, y_test, verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.5620 - accuracy: 0.7406 - 480ms/epoch - 2ms/step
Loss: 0.5620306134223938, Accuracy: 0.7406414151191711


Reults of test 3:

Application cutoff = 500
Classification cutoff = 900
Layer1 = 50
Layer2 = 20
Loss = 0.562
Accuracy = 0.740

Attempt 4

In [38]:
number_input_features = len(X_train_scaled[0])
hidden_nodes_layer1 = 80
hidden_nodes_layer2 = 40

nn_3 = tf.keras.models.Sequential()

nn_3.add(tf.keras.layers.Dense(units = hidden_nodes_layer1,
          input_dim = number_input_features, activation = "tanh"))

nn_3.add(tf.keras.layers.Dense(
    units = hidden_nodes_layer2, activation = "relu"))

nn_3.add(tf.keras.layers.Dense(units = 1, activation = "relu"))

nn_3.summary()

Model: "sequential_4"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_12 (Dense)            (None, 80)                3520      
                                                                 
 dense_13 (Dense)            (None, 40)                3240      
                                                                 
 dense_14 (Dense)            (None, 1)                 41        
                                                                 
Total params: 6,801
Trainable params: 6,801
Non-trainable params: 0
_________________________________________________________________


In [39]:
nn_3.compile(loss = "binary_crossentropy", optimizer = "adam", metrics = ["accuracy"])

In [40]:
fit_model = nn_3.fit(X_train_scaled, y_train, epochs = 100)

Epoch 1/100
804/804 [==============================] - 3s 3ms/step - loss: 2.0454 - accuracy: 0.6655
Epoch 2/100
804/804 [==============================] - 2s 3ms/step - loss: 2.1529 - accuracy: 0.6844
Epoch 3/100
804/804 [==============================] - 2s 2ms/step - loss: 2.3062 - accuracy: 0.6461
Epoch 4/100
804/804 [==============================] - 2s 2ms/step - loss: 1.7971 - accuracy: 0.6709
Epoch 5/100
804/804 [==============================] - 2s 2ms/step - loss: 1.8274 - accuracy: 0.6528
Epoch 6/100
804/804 [==============================] - 3s 3ms/step - loss: 1.5765 - accuracy: 0.6754
Epoch 7/100
804/804 [==============================] - 4s 5ms/step - loss: 3.2296 - accuracy: 0.6117
Epoch 8/100
804/804 [==============================] - 3s 4ms/step - loss: 2.7852 - accuracy: 0.6419
Epoch 9/100
804/804 [==============================] - 4s 5ms/step - loss: 2.1203 - accuracy: 0.6624
Epoch 10/100
804/804 [==============================] - 4s 4ms/step - loss: 1.9575 - accura

In [41]:
model_loss, model_accuracy = nn_3.evaluate(X_test_scaled, y_test, verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.5859 - accuracy: 0.7348 - 504ms/epoch - 2ms/step
Loss: 0.5859088897705078, Accuracy: 0.734810471534729


Results for test 4:

Application cutoff = 500
Classification cutoff = 900
Layer1 = 80
Layer2 = 40
Loss = 0.58590
Accuracy = 0.734810

In [49]:
# Export our model to HDF5 file
from google.colab import files

nn.save('/MyDrive/AlphabetSoupCharity_Optimization1.h5')
nn_1.save('/MyDrive/AlphabetSoupCharity_Optimization2.h5')
nn_2.save('/MyDrive/AlphabetSoupCharity_Optimization3.h5')
nn_3.save('/MyDrive/AlphabetSoupCharity_Optimization4.h5')